### 정규화(Regularization)
#### 1. What is a good model? Explanatoy modeling + Predictive modeling
##### 1) Explanatoy Modeling : 현재 데이터를 얼마나 잘 설명하는가? > Training Error 최소 == MSE 최소
##### 2) Predictive Modeling : 미래 데이터에 대해 얼마나 잘 예측하는가? Expected Error(Testing Error) 최소

#### 2. Expected MSE  = Irreducible Error + Bias^2(편향) + Variance(변동) 
##### 1) Low Bias, Low Variance / High Bias, Low Variance / Low Bias, High Variance / High Bias, High Variance > Low 일수록 좋다.
##### 2) 과녁의 정중앙은 Test Data에 대한 정답
##### 3) 성능상 1, 4 번은 좋고 않좋음이 분명하지만, 2,3번은 상황에 따라 다르다.

#### 3. BLUE( Best Linear Unbiased Estimator ) : 최소제곱법으로 계산된 회귀계수는 편향되지 않는 점추정량 
##### 1) Biased 관점으로만 해석.
##### 2) Unbiased Estimator 중 Variance가 가장 작은 모델을 선택하는 과정

#### 4. Unbiased를 약간 희생하더라도(약간 Biased 하게 해도) Variance(분산)을 감소시키면 미래에 대한 예측이 더 좋아지지 않을까?
##### 1) SubSet Selection : 전체 p개의 설명변수(X) 중 일부 k개만 사용하여 회귀계수 B추정
##### 2) 1차, 2차, 4차식에 대한 차수(X축)/오류(Y축) 
##### 3) 1차일 때는 Training Data와 Testing Data 오류가 높다. (High Bias, Low Variance), 2번(과녁의 정중앙 옆에 모여)
##### 4) 4차일 때는 Training Data의 오류는 낮고, Testing Data의 오류는 다시 높아진다. (Low Bias, High Variance), 3번 (과녁의 정중앙 방향이지만 퍼져)
##### 5) 2차식은 Bias를 약간 희생하고 Variance를 감소시키기 때문에 현재 데이터 설명력뿐만 아니라 미래 데이터 예측도 적합한 모델

#### 5. 정규화 : Least Square Solution으로 Training Data에 overfitting(과적합) 모델을 만들었을 때  B(회귀계수)에 제약을 주어 정돈한다는 의미
#### 6. Ridge Regression
##### 1) B(회귀계수)의 L2(제곱) normalization을 제약 
##### 2) 추정된 릿지 B값이 최소가 되기 위해? 람다가 커지면 B값(B의 영향력)작아지고, 람다가 작아지면 B값(B의 영향력)이 커진다.(0에 가까워져야 한다.)
##### 3) 람다가 커진다? (is very Big) : B를 크게 제약하며, (Training Data) High Biased, Low Variance > 현재 데이터에 대한 설명력이 떨어진다.(UnderFitting)
##### 4) 람다가 작아진다? (is very Small) : B의 제약이 거의 없다, (Testing Data) Low Biased, High Variance > 미래 데이터에 대한 예측력이 떨어진다(OverFitting)
##### 5) 람다(Hyper Parameter, 사용자 결정)을 통해 3), 4)의 상충관계를 잘 조절 필요
##### 6) Least Square Method : B가 무슨 값을 가져도 좋으니 MSE 값만 최소화(Biased 관점)
##### 7) Regulation Method : MSE + Variance 고려, LSM보다는 Biased 하겠지만 미래 데이터에 대한 예측력은 향상
##### 8) B1, B2의 제약조건(원) + MSE 타원 그림(2->3), penalty(U 형태), closed form solution

#### 7. LASSO(Least Absolute 'Shrinkage' and 'Selection' Operator)
##### 1) Shrinkage(회귀계수 파라미터가 값이 작아진다), Selection(Y값 예측에 중요하지 않은 변수 선택을 통해 중요한 변수를 알 수 있다.)
##### 2) SUM(|B|) <=  T, 람다가 커지면 (T값이 작아지면) B에 대한 제약을 크게하며, 람다가 작아지면(T값이 커지면) B에 대한 제약을 거의 없게 한다.
##### > 람다가 커질수록 B값들이 작아진다(Shrinkage, 0으로 수렴)
##### > 람다가 커진다? : 점점 Least Square Solution에 가까워지며, 적은 변수, 간단한 모델, 해석 쉬움, 높은 학습 오차(Training Data에 대한 Under Fitting)
##### > 람다가 작아진다? : 많은 변수, 복잡한 모델, 해석 어려움, 낮은 학습 오차(Training Data에 대한 Overfitting)
##### 3) B1, B2의 제약조건(마름모) + MSE 타원 그림(2->3), penalty(V 형태), numerical Optimization (closed form solution X)
##### 4) 마름모이기 때문에 꼭지점이 존재하며, 이는 Y를 예측하는데 중요하지 않는 변수로 판단.

#### 8. Robust Model(강건한 모델)
##### 1) 데이터가 달라질 때마다 변수 선택이 달라지지 않을까? 변화가 작다면 Robust Model(강건한 모델)이라고 할 수 있다.
##### 2) LASSO의 경우 B값들의 박스플롯 범위가 작다 > Robust Model이라고 할 수 있다.

#### 9. Solution Path
##### 1) Ridge : Shrinkage(람다값이 커질수록, T값이 작아질 수록 B값은 작아진다) vs  LASSO : Shrinkage + Selection
##### 2) Ridge : B값이 작아지지만 0은 되지 않는다. 변수 간 상관관계 높은 상황에서 예측 성능이 좋다.
##### 3) LASSO : B값이 0이 될 수 있다. 변수 간 상관관계가 높은 상황에서 Ridge보다 상대적으로 예측 성능이 떨어진다.
###### > LASSO의 한계점 변수 간 상관관계가 높을 경우 B값의 박스플롯의 범위가 넒어진다.

#### 10. Elastic Modeling
##### 1) LASSO 모델에 Ridge 추가 제약
##### 2) Elastic으로 추정된 B1과 B2의 상관관계가 1이면 이는 Y값을 예측하는데 동일하게 중요한/중요하지 않는 변수이다. (Grouping Effect)
##### 3) Grid Search를 통해 람다1, 람다2를 찾아서 에러가 가장 작은 B값들을 추정
##### 4) B1, B2의 제약조건(원X, 마름모X) + MSE 타원 그림(2->3), penalty(Ridge와 Lasso의 중간 형태)

#### 11. MAE(Mean Absolute Error) vs RMSE(Root Mean Squared Error)
##### 1) MAE : 현실적이고 직관적인 에러
##### 2) RMSE : 크기에 의존적인 에러
##### 3) 매우 큰 에러로 오분류가 된 데이터가 들어오면 RMSE가 해당 데이터에 큰 패널티를 부여
##### 4) 예측 모델이 데이터를 학습할 때 에러가 큰 데이터(특이값)에 민감하지 않고 잘 학습하기 위함(Scaled-Dependet Error, 에러에 대한 Robust한 Metric)

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#11.Ridge,-Lasso,-Elastic-선형회귀분석-과제" data-toc-modified-id="11.Ridge,-Lasso,-Elastic-선형회귀분석-과제-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>11.Ridge, Lasso, Elastic 선형회귀분석 과제</a></span><ul class="toc-item"><li><span><a href="#전처리-:--id,-date-삭제" data-toc-modified-id="전처리-:--id,-date-삭제-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>전처리 :  id, date 삭제</a></span></li><li><span><a href="#범주형-변수-전처리" data-toc-modified-id="범주형-변수-전처리-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>범주형 변수 전처리</a></span></li><li><span><a href="#종속변수-y-=-price-,-나머지는-독립변수-X로-설정" data-toc-modified-id="종속변수-y-=-price-,-나머지는-독립변수-X로-설정-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>종속변수 y = price , 나머지는 독립변수 X로 설정</a></span></li><li><span><a href="#홀드아웃-기법을-사용하여-데이터-분할" data-toc-modified-id="홀드아웃-기법을-사용하여-데이터-분할-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>홀드아웃 기법을 사용하여 데이터 분할</a></span></li><li><span><a href="#Ridge-회귀분석을-하여-R2-값과-RMSE값을-구하고-설명하세요." data-toc-modified-id="Ridge-회귀분석을-하여-R2-값과-RMSE값을-구하고-설명하세요.-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Ridge 회귀분석을 하여 R2 값과 RMSE값을 구하고 설명하세요.</a></span></li><li><span><a href="#스케일링-한-데이터셋의-결과와-스케일링-하지-않은-데이터-셋의-결과를-비교해보세요." data-toc-modified-id="스케일링-한-데이터셋의-결과와-스케일링-하지-않은-데이터-셋의-결과를-비교해보세요.-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>스케일링 한 데이터셋의 결과와 스케일링 하지 않은 데이터 셋의 결과를 비교해보세요.</a></span></li><li><span><a href="#회귀-계수-중-집값과-상관성이-높은-3가지-변수를-해석하세요" data-toc-modified-id="회귀-계수-중-집값과-상관성이-높은-3가지-변수를-해석하세요-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>회귀 계수 중 집값과 상관성이 높은 3가지 변수를 해석하세요</a></span></li></ul></li></ul></div>

# 11.Ridge, Lasso, Elastic 선형회귀분석 과제 

변수 설명 

id - 판매된 각 주택의 고유 ID <br>
date  - 주택 판매 날짜<br>
price - 판매된 각 주택의 가격<br>
bedrooms  - 침실 수<br>
bathrooms  - 화장실이 있지만 샤워 시설은 없는 방을 0.5로 설명하는 욕실 수<br>
sqft_living - 아파트 내부 생활 공간의 평방 피트<br>
sqft_lot - 토지 공간의 평방 피트수<br>
floors  - 층 수<br>
waterfront - 아파트가 해안가를 내려다보고 있는지 여부에 대한 더미 변수<br>
view - 부동산에 대한 뷰가 얼마나 좋았는지에 대한 0에서 4 사이의 인덱스<br>
condition - 아파트 상태에 대한 1에서 5까지의 지수,<br>
grade - 1-13까지의 수로,1-3은 건축 및 설계 미달, 7은 시공 및 설계 평균, 11-13은 시공 및 설계 수준 높은 수준이다.<br>
sqft_above - 지면보다 높은 내부 주택 공간의 평방 피트수<br>
sqft_basement - 지면 아래에 있는 내부 주택 공간의 평방 피트수<br>
yr_built - 집이 처음 지어진 연도<br>
yr_renovated - 집의 마지막 리모델링 연도<br>
sqft_living15 - 가장 가까운 이웃 15명을 위한 내부 주택 생활 공간의 평방 피트<br>
sqft_lot15 - 가장 가까운 15개 이웃 토지 부지의 평방 피트수<br>

In [91]:
import pandas as pd 
df = pd.read_csv("data/kc_house_data.csv")
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors waterfront  view  condition  grade  \
0             1180      5650     1.0   standard     0          3      7   
1             2570      7242     2.0   standard     0          3      7   
2              770     10000     1.0   standard     0          3      6   
3             1960      5000     1.0   standard     0          5      7   
4             1680      8080     1.0   standard     0          3      8   
...            ...       ...     ...        ...   ...        ...    ...   
21608         1530      1131     3.0   standard     0          3      8   
21609         2310      5813     2.0   standard     0          3      8   
21610         1020      1350     2.0   standard     0          3      7   
21611         1600      2388     2.0   standard     0          3      8   
21612         1020      1076     2.0   standard     0          3      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  sqft_living15  \
0            1180              0      1955             0           1340   
1            2170            400      1951          1991           1690   
2             770              0      1933             0           2720   
3            1050            910      1965             0           1360   
4            1680              0      1987             0           1800   
...           ...            ...       ...           ...            ...   
21608        1530              0      2009             0           1530   
21609        2310              0      2014             0           1830   
21610        1020              0      2009             0           1020   
21611        1600              0      2004             0           1410   
21612        1020              0      2008             0           1020   

       sqft_lot15  
0            5650  
1            7639  
2            8062  
3            5000  
4            7503  
...           ...  
21608        1509  
21609        7200  
21610        2007  
21611        1287  
21612        1357  

[21613 rows x 18 columns]

In [92]:
# 1. 전처리
# 1-1) 결측치 확인 
df.info() # 결측치 X

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  object 
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  sqft_living15  21613 non-null  int64  
 17  sqft_lot15     21613 non-null  int64  
dtypes: flo

In [93]:
# 1-2) id, date 삭제
df.drop(columns = ['id', 'date'], inplace = True)
df

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0      221900.0         3       1.00         1180      5650     1.0   
1      538000.0         3       2.25         2570      7242     2.0   
2      180000.0         2       1.00          770     10000     1.0   
3      604000.0         4       3.00         1960      5000     1.0   
4      510000.0         3       2.00         1680      8080     1.0   
...         ...       ...        ...          ...       ...     ...   
21608  360000.0         3       2.50         1530      1131     3.0   
21609  400000.0         4       2.50         2310      5813     2.0   
21610  402101.0         2       0.75         1020      1350     2.0   
21611  400000.0         3       2.50         1600      2388     2.0   
21612  325000.0         2       0.75         1020      1076     2.0   

      waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0       standard     0          3      7        1180              0      1955   
1       standard     0          3      7        2170            400      1951   
2       standard     0          3      6         770              0      1933   
3       standard     0          5      7        1050            910      1965   
4       standard     0          3      8        1680              0      1987   
...          ...   ...        ...    ...         ...            ...       ...   
21608   standard     0          3      8        1530              0      2009   
21609   standard     0          3      8        2310              0      2014   
21610   standard     0          3      7        1020              0      2009   
21611   standard     0          3      8        1600              0      2004   
21612   standard     0          3      7        1020              0      2008   

       yr_renovated  sqft_living15  sqft_lot15  
0                 0           1340        5650  
1              1991           1690        7639  
2                 0           2720        8062  
3                 0           1360        5000  
4                 0           1800        7503  
...             ...            ...         ...  
21608             0           1530        1509  
21609             0           1830        7200  
21610             0           1020        2007  
21611             0           1410        1287  
21612             0           1020        1357  

[21613 rows x 16 columns]

In [94]:
# 1-3) 범주형 변수 전처리
print(df['waterfront'].value_counts())
df = pd.get_dummies(df, columns = ['waterfront', 'view', 'condition','grade'])
df

standard      21450
river_view      163
Name: waterfront, dtype: int64


price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0      221900.0         3       1.00         1180      5650     1.0   
1      538000.0         3       2.25         2570      7242     2.0   
2      180000.0         2       1.00          770     10000     1.0   
3      604000.0         4       3.00         1960      5000     1.0   
4      510000.0         3       2.00         1680      8080     1.0   
...         ...       ...        ...          ...       ...     ...   
21608  360000.0         3       2.50         1530      1131     3.0   
21609  400000.0         4       2.50         2310      5813     2.0   
21610  402101.0         2       0.75         1020      1350     2.0   
21611  400000.0         3       2.50         1600      2388     2.0   
21612  325000.0         2       0.75         1020      1076     2.0   

       sqft_above  sqft_basement  yr_built  yr_renovated  ...  grade_4  \
0            1180              0      1955             0  ...        0   
1            2170            400      1951          1991  ...        0   
2             770              0      1933             0  ...        0   
3            1050            910      1965             0  ...        0   
4            1680              0      1987             0  ...        0   
...           ...            ...       ...           ...  ...      ...   
21608        1530              0      2009             0  ...        0   
21609        2310              0      2014             0  ...        0   
21610        1020              0      2009             0  ...        0   
21611        1600              0      2004             0  ...        0   
21612        1020              0      2008             0  ...        0   

       grade_5  grade_6  grade_7  grade_8  grade_9  grade_10  grade_11  \
0            0        0        1        0        0         0         0   
1            0        0        1        0        0         0         0   
2            0        1        0        0        0         0         0   
3            0        0        1        0        0         0         0   
4            0        0        0        1        0         0         0   
...        ...      ...      ...      ...      ...       ...       ...   
21608        0        0        0        1        0         0         0   
21609        0        0        0        1        0         0         0   
21610        0        0        1        0        0         0         0   
21611        0        0        0        1        0         0         0   
21612        0        0        1        0        0         0         0   

       grade_12  grade_13  
0             0         0  
1             0         0  
2             0         0  
3             0         0  
4             0         0  
...         ...       ...  
21608         0         0  
21609         0         0  
21610         0         0  
21611         0         0  
21612         0         0  

[21613 rows x 36 columns]

In [95]:
# 1-4) 종속변수 y = price, 나머지는 독립변수 X로 설정
y = df['price']
X = df.drop(columns = ['price'])

In [96]:
# 1-5) 홀드아웃 기법을 사용하여 데이터 분할
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 1000)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(15129, 35)
(15129,)
(6484, 35)
(6484,)


In [97]:
# 2. Ridge 회귀분석
import warnings
warnings.filterwarnings(action = 'ignore')

from sklearn.linear_model import Ridge
import numpy as np

ridge = Ridge()
ridge.fit(X_train, y_train)

print("사용한 컬럼의 수 : " , np.sum(ridge.coef_ != 0)) # Ridge의 계수, 0이 아니면 해당 컬럼을 사용
print("Train 데이터 성능 : ", ridge.score(X_train, y_train))
print("Test 데이터 성능 : ", ridge.score(X_test, y_test))

사용한 컬럼의 수 :  35
Train 데이터 성능 :  0.6882164616743431
Test 데이터 성능 :  0.6696311193094535


In [98]:
# 2-1) RMSE
y_pred = ridge.predict(X_test)
resid = y_test - y_pred
rmse = np.sqrt(sum(resid**2)/len(y_test))
print("RMSE : " , rmse) # 

from sklearn.metrics import mean_squared_error
print("RMSE : " , np.sqrt(mean_squared_error(y_test, y_pred)))

RMSE :  203440.4717711172
RMSE :  203440.47177111704


##### Before Scaling
##### Train Data는 68%, Test Data는 66% 이므로 Overfitting
##### Test Data의 66% 설명력은 좋은 성능이 아니라고 할 수 있으며, RMSE를 통해 예측 집값과 실제 집값은 평균 203,440 만큼 차이가 발생한다. 

In [99]:
# 2-2) GridSearch 추가
from sklearn.model_selection import GridSearchCV
ridge = Ridge()
param_grid = {'alpha' : (0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.5, 1, 2, 3)}
model = GridSearchCV(ridge, param_grid)
model.fit(X_train, y_train)

GridSearchCV(estimator=Ridge(),
             param_grid={'alpha': (0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05,
                                   0.1, 0.15, 0.2, 0.25, 0.3, 0.5, 1, 2, 3)})

In [100]:
model.best_estimator_

Ridge(alpha=1)

In [101]:
ridge = Ridge(alpha = 3)
ridge.fit(X_train, y_train)

Ridge(alpha=3)

In [102]:
print("사용한 컬럼의 수 : " , np.sum(ridge.coef_ != 0)) # Ridge의 계수, 0이 아니면 해당 컬럼을 사용
print("Train 데이터 성능 : ", ridge.score(X_train, y_train))
print("Test 데이터 성능 : ", ridge.score(X_test, y_test))

사용한 컬럼의 수 :  35
Train 데이터 성능 :  0.6871903930681482
Test 데이터 성능 :  0.669037508955745


##### Before Scaling + GridSearch > Before Scaling 동일
##### Train Data는 68%, Test Data는 66% 이므로 Overfitting
##### Test Data의 66% 설명력은 좋은 성능이 아니라고 할 수 있으며, RMSE를 통해 예측 집값과 실제 집값은 평균 203,440 만큼 차이가 발생한다. 

In [103]:
# 3. 스케일링 vs 스케일링 X
# Tip : X_train 데이터를 fit_transform 하여 전처리, X_test는 transform 사용하여 전처리 <br>
#  X_train_scaled = pd.DataFrame(scaler.fit_transform(X_train),columns=X_train.columns,index=X_train.index)

from sklearn.preprocessing import StandardScaler # 표준화 : 각 컬럼을 표준정규분포(0,1)로 Scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

ridge = Ridge()
ridge.fit(X_train_scaled, y_train)
y_pred_scaled = ridge.predict(X_test_scaled)

In [104]:
from sklearn.metrics import mean_squared_error
print("사용한 컬럼의 수 : " , np.sum(ridge.coef_ != 0)) # Ridge의 계수, 0이 아니면 해당 컬럼을 사용
print("Train 데이터 성능 : ", ridge.score(X_train_scaled, y_train))
print("Test 데이터 성능 : ",  ridge.score(X_test_scaled, y_test))
print("RMSE : ",  np.sqrt(mean_squared_error(y_test, y_pred))) 

사용한 컬럼의 수 :  35
Train 데이터 성능 :  0.6884224825516956
Test 데이터 성능 :  0.6697700336650881
RMSE :  203440.47177111704


##### StandardScaler()를 진행을 해도 StandardScaler() 전과 동일하다

In [105]:
# 4. 회귀 계수 중 집값과 상관성이 높은 3가지 변수를 해석

In [106]:
df_coef = pd.DataFrame(ridge.coef_, index = X.columns, columns = ['coef'])
df_coef['abs_coef'] = abs(df_coef['coef'])
df_coef.sort_values('abs_coef', ascending = False)

coef      abs_coef
yr_built              -92723.982411  92723.982411
grade_11               72023.462087  72023.462087
grade_10               70111.029295  70111.029295
grade_12               67417.770416  67417.770416
grade_13               64504.297300  64504.297300
sqft_living            60176.119134  60176.119134
grade_6               -50634.064559  50634.064559
grade_7               -47987.099282  47987.099282
sqft_above             44792.060929  44792.060929
sqft_basement          41174.532385  41174.532385
grade_9                39183.186163  39183.186163
bathrooms              37296.823020  37296.823020
bedrooms              -25192.412559  25192.412559
floors                 24420.491799  24420.491799
view_4                 23518.787806  23518.787806
waterfront_standard   -21812.266109  21812.266109
waterfront_river_view  21812.266109  21812.266109
grade_5               -21750.537563  21750.537563
sqft_living15          20410.834289  20410.834289
view_0                -16268.888994  16268.888994
sqft_lot15            -14478.479086  14478.479086
condition_5            11288.587571  11288.587571
grade_8                -8069.149070   8069.149070
view_1                  7954.739566   7954.739566
view_3                  7729.169856   7729.169856
condition_3            -7139.790482   7139.790482
grade_4                -6062.321455   6062.321455
yr_renovated            5740.636709   5740.636709
condition_2            -3505.303228   3505.303228
grade_3                -2165.928602   2165.928602
condition_4             1736.854314   1736.854314
condition_1            -1428.857067   1428.857067
grade_1                 -977.371339    977.371339
view_2                  -926.533257    926.533257
sqft_lot                 -45.389789     45.389789

##### 가장 상관관계가 높은 변수는 yr_built, grade(10~13), sqft_living이다.

In [107]:
# 2. LASSO 회귀분석
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error
import numpy as np
lasso = Lasso()
param_grid = {'alpha' : (0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.5, 1, 2, 3)}
model = GridSearchCV(lasso, param_grid)
model.fit(X_train, y_train)
print("최적의 알파값 : " , model.best_estimator_)

최적의 알파값 :  Lasso(alpha=3)


In [108]:
lasso = Lasso(alpha = 3)
lasso.fit(X_train, y_train)
y_pred = lasso.predict(X_test)

print('사용한 컬럼의 수 : ', np.sum(lasso.coef_ !=0))
print('Train 데이터의 성능(R^2) : ', lasso.score(X_train, y_train))
print('Test 데이터의 성능(R^2) : ', lasso.score(X_test, y_test))
print('RMSE : ', np.sqrt(mean_squared_error(y_test, y_pred)))

사용한 컬럼의 수 :  34
Train 데이터의 성능(R^2) :  0.6884209100914462
Test 데이터의 성능(R^2) :  0.6697566120065335
RMSE :  203401.82902975834


In [110]:
# ElasticNet 회귀분석
from sklearn.linear_model import ElasticNet
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error
import numpy as np

elasticNet = ElasticNet()
param_grid = {'alpha' : (0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.5, 1, 2, 3)
              ,'l1_ratio' : [0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8, 0.9]} # LASSO(L1) + Ridge(L2)
model = GridSearchCV(elasticNet, param_grid)
model.fit(X_train, y_train)
print('최적의 알파값 : ', model.best_estimator_)

최적의 알파값 :  ElasticNet(alpha=0.0001, l1_ratio=0.3)


In [111]:
elasticNet = ElasticNet(alpha = 0.0001, l1_ratio=0.3)
elasticNet.fit(X_train,y_train)
y_pred = elasticNet.predict(X_test)

print('사용한 컬럼의 수 : ', np.sum(elasticNet.coef_ !=0))
print('Train 데이터의 성능(R^2) : ', elasticNet.score(X_train, y_train))
print('Test 데이터의 성능(R^2) : ', elasticNet.score(X_test, y_test))
print('RMSE : ', np.sqrt(mean_squared_error(y_test, y_pred)))

사용한 컬럼의 수 :  35
Train 데이터의 성능(R^2) :  0.6881951652749771
Test 데이터의 성능(R^2) :  0.6696146073964215
RMSE :  203445.5557084991
